{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/shreyansegnyte/NASA-GeneLab-Code/blob/main/5_quantifying_gene_expression.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<div>\n",
        "<img src=\"https://www.nasa.gov/wp-content/uploads/2024/07/osdr-gl4hs-logo.png\" width=\"600\"/>\n",
        "</div>"
      ],
      "metadata": {
        "id": "heuFk2XLbAg8"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **NOTEBOOK 5: Quantifying gene expression**\n",
        "In this notebook, you will assign a number for each gene that represents the number of RNA transcripts for that gene that your sample expressed in the tissue at a point in time.\n",
        "\n",
        "## **Objectives of this notebook**\n",
        "The primary objective of this notebook is to quantify the gene expression of the one sample's reduced chromosome 17 alignment. You will then compare the gene expression counts for your sample's chromosome 17 to those obtained by the GeneLab processing team. We expect that the quantities should be off by a factor approximately near `REDUCTION_FACTOR`.\n",
        "\n",
        "## **UNIX commands introduced in this notebook**\n",
        "\n",
        "`grep` command to search for lines in files that have a matching pattern.\n",
        "\n",
        "`htseq-count ` command to quantify gene expression.\n",
        "\n"
      ],
      "metadata": {
        "id": "MbsdHhEvn0cc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prepare your environment for this lab"
      ],
      "metadata": {
        "id": "5oHJQNkXTqed"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wuZaffD5o3mS"
      },
      "outputs": [],
      "source": [
        "# mount your google drive\n",
        "from google.colab import drive\n",
        "drive.flush_and_unmount()\n",
        "drive.mount(\"mnt\")\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "start_time=datetime.datetime.now()\n",
        "print('notebook start time: ', start_time.strftime('%Y-%m-%d %H:%M:%S'))"
      ],
      "metadata": {
        "id": "zCQjiOl9_JqA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set env variables for OSD dataset to use in this lab\n",
        "OSD_DATASET='104'\n",
        "GLDS_DATASET='104'"
      ],
      "metadata": {
        "id": "y0zHUqUXe9J4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set FASTQ_DIR directory location in google drive\n",
        "import os\n",
        "FASTQ_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/FASTQ\"\n",
        "if not os.path.exists(FASTQ_DIR):\n",
        "  raise Exception(\"STOP! You must finish the previous notebooks before running this one\")"
      ],
      "metadata": {
        "id": "09QH8xP-yXcU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# read env var for reduction factor from first notebook\n",
        "import os\n",
        "with open(f\"{FASTQ_DIR}/SAMPLE_NAME.txt\", \"r\") as f:\n",
        "  OSD_SAMPLE=f.read().strip()\n",
        "if not OSD_SAMPLE:\n",
        "  raise Exception(\"STOP! You must finish the previous notebooks before running this one\")\n",
        "print(OSD_SAMPLE)"
      ],
      "metadata": {
        "id": "B3F65jsTSF6n"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# read env var for reduction factor from first notebook\n",
        "import os\n",
        "with open(f\"{FASTQ_DIR}/REDUCTION_FACTOR.txt\", \"r\") as f:\n",
        "  REDUCTION_FACTOR=f.read()\n",
        "if not REDUCTION_FACTOR:\n",
        "  raise Exception(\"STOP! You must finish the previous notebooks before running this one\")\n",
        "print(REDUCTION_FACTOR)"
      ],
      "metadata": {
        "id": "QLuaYNI-l7pp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set REFERENCE_DIR directory location in google drive\n",
        "import os\n",
        "REFERENCE_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/REFERENCE\"\n",
        "if not os.path.exists(REFERENCE_DIR):\n",
        "  raise Exception(\"STOP! You must finish the previous notebooks before running this one\")"
      ],
      "metadata": {
        "id": "fZPjJTKKVrK5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set ALIGNMENT_DIR directory location in google drive\n",
        "import os\n",
        "ALIGNMENT_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/STAR/ALIGNMENT\"\n",
        "if not os.path.exists(ALIGNMENT_DIR):\n",
        "  raise Exception(\"STOP! You must finish the previous notebooks before running this one\")"
      ],
      "metadata": {
        "id": "z4t3vgUmYgMa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set COUNTS_DIR directory location in google drive\n",
        "import os\n",
        "COUNTS_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/COUNTS\"\n",
        "if os.path.exists(COUNTS_DIR):\n",
        "  !rm -rf {COUNTS_DIR}\n",
        "\n",
        "!mkdir -p {COUNTS_DIR}"
      ],
      "metadata": {
        "id": "LlMgkdGsYrJS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# install htseq\n",
        "!pip install HTSeq"
      ],
      "metadata": {
        "id": "mpgYCrDrLRI8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# determine the version of htseq installed\n",
        "!htseq-count --version"
      ],
      "metadata": {
        "id": "-j2BT8klO_by"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# download gene annotation for GRCm39\n",
        "import os\n",
        "if os.path.exists(f\"{REFERENCE_DIR}\"):\n",
        "  !rm -rf {REFERENCE_DIR}\n",
        "!mkdir -p {REFERENCE_DIR}\n",
        "  !wget -O {REFERENCE_DIR}/gencode.vM36.primary_assembly.basic.annotation.gtf.gz \\\n",
        "    https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M36/gencode.vM36.primary_assembly.basic.annotation.gtf.gz"
      ],
      "metadata": {
        "id": "BMqeSDqmOa5p"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# remove all but chr17 annotations\n",
        "# change \"chr17\" to \"17\" in the GTF file as that's what the htseq-count is expecting\n",
        "# gzip the file\n",
        "!gunzip -c {REFERENCE_DIR}/gencode.vM36.primary_assembly.basic.annotation.gtf.gz > {REFERENCE_DIR}/gencode.vM36.primary_assembly.basic.annotation.gtf\n",
        "!grep ^chr17 {REFERENCE_DIR}/gencode.vM36.primary_assembly.basic.annotation.gtf | sed 's/chr17/17/' > {REFERENCE_DIR}/chr17.gtf\n",
        "!gzip -c {REFERENCE_DIR}/chr17.gtf > {REFERENCE_DIR}/chr17.gtf.gz\n"
      ],
      "metadata": {
        "id": "X_end-Xwr5fE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read the [GTF documentation](https://www.gencodegenes.org/mouse/) and the [GTF Wikipedia page](https://en.wikipedia.org/wiki/Gene_transfer_format) for more information about basic mouse gene annotation."
      ],
      "metadata": {
        "id": "voLmgHVs9Hpe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# check the first 10 lines of the GTF annotation file\n",
        "!head -10 {REFERENCE_DIR}/chr17.gtf"
      ],
      "metadata": {
        "id": "IGFUld9jZmNw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Question: What is lncRNA? Feel free to read more about that in this [Wikipedia article](https://en.wikipedia.org/wiki/Long_non-coding_RNA)."
      ],
      "metadata": {
        "id": "zij_0NUf-qeR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Use HTSEQ to quantify gene expression"
      ],
      "metadata": {
        "id": "NQ1clO-WTyR8"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read the [htseq-count manual](https://htseq.readthedocs.io/en/master/htseqcount.html) for more information."
      ],
      "metadata": {
        "id": "R9OLuFnD8xEV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# run htseq to quantify the gene expression\n",
        "\n",
        "!htseq-count -n 2 \\\n",
        "  --format bam \\\n",
        "  --order pos \\\n",
        "  --stranded reverse \\\n",
        "  {ALIGNMENT_DIR}/chr17Aligned.out.bam \\\n",
        "  {REFERENCE_DIR}/chr17.gtf.gz \\\n",
        "  > {COUNTS_DIR}/chr17-counts.tsv"
      ],
      "metadata": {
        "id": "brTAHVZxLSpk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Note that you may get a warning about \"mate records missing\" from `htseq-count`. You can ignore this warning -- it's a known bug in the `htseq-count` software. Read [this github issue](https://github.com/simon-anders/htseq/issues/37) for more information if you're curious."
      ],
      "metadata": {
        "id": "EPSbpsRCyflm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# look at the first 10 lines of the counts file\n",
        "!head -10 {COUNTS_DIR}/chr17-counts.tsv"
      ],
      "metadata": {
        "id": "4Lgg7ABl67hm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# read count data from file into dataframe\n",
        "import pandas as pd\n",
        "counts_df=pd.read_csv(f\"{COUNTS_DIR}/chr17-counts.tsv\", sep=\"\\t\", header=None)\n",
        "counts_df.head()"
      ],
      "metadata": {
        "id": "QY8_lbHSgtEH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# remove any rows from the counts_df that do not begin with 'ENSMUSG'\n",
        "print('length before filter: ', len(counts_df))\n",
        "counts_df=counts_df[counts_df[0].str.startswith('ENSMUSG')]\n",
        "print('length after filter: ', len(counts_df))\n",
        "counts_df[0]"
      ],
      "metadata": {
        "id": "rdO5v1VnkUPb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Compare your count data to the GeneLab-processed count data for the same sample"
      ],
      "metadata": {
        "id": "J0JNMn114uZu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# open another tab in your web browser and navigate to the following site:\n",
        "url=!echo https://visualization.osdr.nasa.gov/biodata/api/v2/dataset/OSD-{OSD_DATASET}/files/\\?format=browser\n",
        "print(url[0])"
      ],
      "metadata": {
        "id": "ap8hqs_jfZbH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# download genelab-processed data for OSD dataset\n",
        "import pandas as pd\n",
        "#url = 'https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + OSD_DATASET + '/download?source=datamanager\\&file=GLDS-' + GLDS_DATASET + '_rna_seq_STAR_Unnormalized_Counts.csv'url='https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + osd_dataset + '/download?source=datamanager\\&file=GLDS-' + glds_dataset + '_rna_seq_STAR_Unnormalized_Counts.csv'\n",
        "url = 'https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + OSD_DATASET + '/download?source=datamanager\\&file=GLDS-' + GLDS_DATASET + '_rna_seq_Unnormalized_Counts.csv'\n",
        "osd_df = pd.read_csv(url)\n",
        "osd_df.head()"
      ],
      "metadata": {
        "id": "xvJrFiof7SN9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# get list of genes from OSD_SAMPLE counts data to compare with the counts for the entire OSD_DATASET\n",
        "# remove the \".\" extension from the ensemble gene ID\n",
        "sample_genes = list(counts_df[0].values)[:20]\n",
        "sample_genes = [gene.split(\".\")[0] for gene in sample_genes]\n",
        "sample_genes"
      ],
      "metadata": {
        "id": "DkpQkZKXnlwm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# find the gene count data for the first 20 genes associated with the sample name\n",
        "osd_df[['Unnamed: 0', OSD_SAMPLE]].head(20)\n",
        "gene_counts_from_osd = list(osd_df[osd_df['Unnamed: 0'].isin(sample_genes)][OSD_SAMPLE].values)"
      ],
      "metadata": {
        "id": "t2FRqfLcQnrB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# capture the first 20 lines of the counts_df dataframe\n",
        "gene_counts_from_you = list(counts_df[1].values)[:20]"
      ],
      "metadata": {
        "id": "Wen-DLnVtwMB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# compare the counts side by side (the second column should be roughly 1/REDUCTION_FACTOR of the first column)\n",
        "for gene_count in zip(sample_genes, gene_counts_from_osd, gene_counts_from_you):\n",
        "  print(gene_count[0], '\\t', gene_count[1], '\\t', gene_count[2])\n"
      ],
      "metadata": {
        "id": "feX872BruWMs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# look at the first 500 gene counts in both\n",
        "# determine if fraction of abundance is approximately 1/REDUCTION_FACTOR\n",
        "import numpy as np\n",
        "count_fractions = list()\n",
        "for gene in counts_df[0].values[:500]:\n",
        "  _gene = gene.split(\".\")[0]\n",
        "  if _gene in osd_df['Unnamed: 0'].values:\n",
        "    genelab_val = osd_df[osd_df['Unnamed: 0'] == _gene][OSD_SAMPLE].values[0]\n",
        "    your_val = int(counts_df[counts_df[0] == gene][1].values[0])\n",
        "    if not genelab_val == 0:\n",
        "      frac = your_val/genelab_val\n",
        "      count_fractions.append(frac)\n",
        "\n",
        "print(np.mean(count_fractions))"
      ],
      "metadata": {
        "id": "trnsLHj9Vr7r"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Check your work before moving on"
      ],
      "metadata": {
        "id": "dy2dyc5aT7sz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# check disk space utilization in google drive (should be about 2.4GB)\n",
        "!du -sh /content/mnt/MyDrive/NASA/GL4HS"
      ],
      "metadata": {
        "id": "1eTNc4-fT-gz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "end_time=datetime.datetime.now()\n",
        "print('notebook end time: ', end_time.strftime('%Y-%m-%d %H:%M:%S'))\n",
        "\n",
        "print('notebook runtime: ', end_time - start_time)\n",
        "#"
      ],
      "metadata": {
        "id": "C4JOAMxU_Mc2"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}