{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/shreyansegnyte/NASA-GeneLab-Code/blob/main/6_performing_dgea.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<div>\n",
        "<img src=\"https://www.nasa.gov/wp-content/uploads/2024/07/osdr-gl4hs-logo.png\" width=\"600\"/>\n",
        "</div>"
      ],
      "metadata": {
        "id": "d0_MkJdDbGIM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **NOTEBOOK 6: Performing differential gene expression analysis**\n",
        "In this notebook, you will run statistical tests to determine which genes are differentially expressed between ground control and space flight groups.\n",
        "\n",
        "## **Objectives of this notebook**\n",
        "The primary objective of this notebook is to perform [differential gene expression analysis](https://www.youtube.com/watch?v=tlf6wYJrwKY) (DGEA) using [Python's implementation of DESeq2](https://pydeseq2.readthedocs.io/en/stable/auto_examples/plot_step_by_step.html#sphx-glr-auto-examples-plot-step-by-step-py) to find genes which have a large enough [log2 fold change](https://www.biostars.org/p/9505003/) between ground control and space flight as well as a low enough adjusted [p-value](https://www.youtube.com/watch?v=vemZtEM63GY) as to be statistically significant.\n",
        "\n",
        "DGEA is actually a series of several [statistical hypothesis tests](https://www.youtube.com/watch?v=0oc49DyA3hU) - one per gene. The null hypothesis is that there is no significant difference between the expression of that gene between ground control and spaceflight groups. The alternate hypothesis is that there is a significant difference between the expression of that gene between those 2 groups. This difference can either be over-regulated gene expression or under-regulated gene expression, making this a [two-tailed test](https://www.youtube.com/watch?v=bsZGt-caXO4).\n",
        "\n",
        "Additionally, you will use principal components analysis (PCA) to visualize the gene expression of all the samples in OSD-104 to determine if the samples separate according to their condition (space flight and ground control). You can learn more about PCA [here](https://youtu.be/HMOI_lkzW08?si=zTSi-OclCvmEvdVR). After performing DGEA, you will create a volcano plot of the p-values to visualize their distribution. You can learn more about volcano plots [here](https://www.youtube.com/watch?v=7aWAdw2jhj0).\n",
        "\n",
        "We will no longer use the FASTQ data from the previous notebooks regarding the single sample paired end reads. Instead, we will download the GeneLab-processed count data for all the samples in OSD-104 and use that as our starting point for this and the next notebook.\n",
        "\n",
        "## **UNIX commands introduced in this notebook**\n",
        "\n",
        "[`unzip`](https://linux.die.net/man/1/unzip) command to extract the files in a ZIP archive."
      ],
      "metadata": {
        "id": "Ftcuq8HIqDcv"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prepare runtime environment for lab"
      ],
      "metadata": {
        "id": "MbsdHhEvn0cc"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wuZaffD5o3mS",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 356
        },
        "outputId": "fdbc63f6-6cb0-4cc7-ce7a-f2e9014e3406"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive not mounted, so nothing to flush and unmount.\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "mount failed",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-1180343536.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mgoogle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolab\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mdrive\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mdrive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush_and_unmount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mdrive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"mnt\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/google/colab/drive.py\u001b[0m in \u001b[0;36mmount\u001b[0;34m(mountpoint, force_remount, timeout_ms, readonly)\u001b[0m\n\u001b[1;32m     95\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmountpoint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mforce_remount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m120000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreadonly\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m   \u001b[0;34m\"\"\"Mount your Google Drive at the specified mountpoint path.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m   return _mount(\n\u001b[0m\u001b[1;32m     98\u001b[0m       \u001b[0mmountpoint\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m       \u001b[0mforce_remount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mforce_remount\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/google/colab/drive.py\u001b[0m in \u001b[0;36m_mount\u001b[0;34m(mountpoint, force_remount, timeout_ms, ephemeral, readonly)\u001b[0m\n\u001b[1;32m    270\u001b[0m             \u001b[0;34m'https://research.google.com/colaboratory/faq.html#drive-timeout'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    271\u001b[0m         )\n\u001b[0;32m--> 272\u001b[0;31m       \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'mount failed'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mextra_reason\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    273\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mcase\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    274\u001b[0m       \u001b[0;31m# Terminate the DriveFS binary before killing bash.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: mount failed"
          ]
        }
      ],
      "source": [
        "# mount google drive to mnt mount point\n",
        "from google.colab import drive\n",
        "drive.flush_and_unmount()\n",
        "drive.mount(\"mnt\")\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "start_time = datetime.datetime.now()\n",
        "print('notebook start time: ', start_time.strftime('%Y-%m-%d %H:%M:%S'))"
      ],
      "metadata": {
        "id": "_7d8Za77MgJS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# define OSD dataset number to use for this notebook\n",
        "\n",
        "OSD_DATASET='104'\n",
        "GLDS_DATASET='104'\n",
        "\n",
        "#OSD_DATASET='511'\n",
        "#GLDS_DATASET='511'\n",
        "\n",
        "#OSD_DATASET='580'\n",
        "#GLDS_DATASET='573'"
      ],
      "metadata": {
        "id": "aYYro7o0n6kY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# define the threshold for coefficient of variation\n",
        "CV_THRESH=0.5"
      ],
      "metadata": {
        "id": "3rTog4Ywgd-2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# create META_DIR directory for metadata\n",
        "import os\n",
        "META_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/META/\" + OSD_DATASET\n",
        "if os.path.exists(META_DIR):\n",
        "  !rm -rf {META_DIR}\n",
        "!mkdir -p {META_DIR}"
      ],
      "metadata": {
        "id": "d-mmPpDq8DXI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# create DGEA_DIR directory for DGEA output\n",
        "DGEA_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/DGEA/\" + OSD_DATASET\n",
        "if os.path.exists(DGEA_DIR):\n",
        "  !rm -rf {DGEA_DIR}\n",
        "\n",
        "!mkdir -p {DGEA_DIR}"
      ],
      "metadata": {
        "id": "16Gxq5i-Qrfs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# install packages and import the modules needed to do DGEA\n",
        "!pip install pyDeseq2 > /dev/null\n",
        "from pydeseq2.dds import DeseqDataSet\n",
        "from pydeseq2.ds import DeseqStats\n",
        "import numpy as np\n",
        "!pip install scikit-learn > /dev/null\n",
        "from sklearn.decomposition import PCA\n",
        "import matplotlib.pyplot as plt\n",
        "!pip install Ensembl_converter > /dev/null\n",
        "from Ensembl_converter import EnsemblConverter"
      ],
      "metadata": {
        "id": "e5Wls0rb8nei"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# define log2 fold change threshold\n",
        "LOG2FC_THRESH = 2"
      ],
      "metadata": {
        "id": "zaUQ8zIKEmDz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Define Python methods for notebook"
      ],
      "metadata": {
        "id": "1wR0LO0fwkuU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# define method for removing NaNs from dataframe\n",
        "def drop_nans(df):\n",
        "  # drop NaN rows\n",
        "  df = df.dropna(inplace=False)\n",
        "  return df"
      ],
      "metadata": {
        "id": "Fpd2_8oSJfq9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can learn more about Python NaN [here](https://stackoverflow.com/questions/17628613/what-is-inf-and-nan)."
      ],
      "metadata": {
        "id": "4WXo35xfGTyi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# define method for dropping low-count genes\n",
        "def drop_lowcount(df, threshold=10):\n",
        "\n",
        "  # let's drop any low-count genes\n",
        "  if 'transcript' in df.columns:\n",
        "    df = df[df.drop(columns=['transcript']).sum(axis=1) >= threshold]\n",
        "  elif 'Unnamed: 0' in df.columns:\n",
        "    df = df[df.drop(columns=['Unnamed: 0']).sum(axis=1) >= threshold]\n",
        "  else:\n",
        "    raise Exception(\"check file format\")\n",
        "  return df\n"
      ],
      "metadata": {
        "id": "6HN5IusVITph"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can read more about low-count gene thresholds [here](https://www.biostars.org/p/251922/)."
      ],
      "metadata": {
        "id": "x0DavLq2GZYO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# define method for filtering low coefficient-of-variation genes\n",
        "def filter_cvs(df, thresh):\n",
        "  import numpy as np\n",
        "  import matplotlib.pyplot as plt\n",
        "\n",
        "  # calculate coefficient of variation\n",
        "  cvs=list()\n",
        "  for i in range(len(df)):\n",
        "    m = np.mean(df.iloc[i][1:])\n",
        "    sd = np.std(df.iloc[i][1:])\n",
        "    cvs.append(sd/m)\n",
        "\n",
        "  # plot hist of dist of coev of variation\n",
        "  fig, axs = plt.subplots()\n",
        "  axs.hist(cvs, bins=20)\n",
        "\n",
        "  # add labels for x and y axes\n",
        "  plt.xlabel('coefficient of variation')\n",
        "  plt.ylabel('number of genes')\n",
        "\n",
        "  # keep genes with cv > thresh\n",
        "  indices = list()\n",
        "  for i in range(len(cvs)):\n",
        "    if cvs[i] > thresh:\n",
        "      indices.append(i)\n",
        "  return df.iloc[indices]\n"
      ],
      "metadata": {
        "id": "kP1XbXTGHwyV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can read more about the coefficient of variation [here](https://en.wikipedia.org/wiki/Coefficient_of_variation)."
      ],
      "metadata": {
        "id": "h9hFDO5LHHFt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# define method to transpose a dataframe\n",
        "def transpose_df(df, cur_index_col, new_index_col):\n",
        "  df = df.set_index(cur_index_col).T\n",
        "  df.reset_index(level=0, inplace=True)\n",
        "  cols = [new_index_col] + list(df.columns)[1:]\n",
        "  df.columns = cols\n",
        "  return df"
      ],
      "metadata": {
        "id": "ZI_Th7520Dlm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can read more about transposition [here](https://www.geeksforgeeks.org/python-pandas-dataframe-transpose/)."
      ],
      "metadata": {
        "id": "me9fQhopHSc1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# define method for dropping ERCC genes from dataframe\n",
        "def drop_ercc(df):\n",
        "  df = df[df['Unnamed: 0'].str.contains('ERCC-') == False]\n",
        "  return df"
      ],
      "metadata": {
        "id": "JRn5oqdm22Eh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can read more about ERCC [here](https://www.nist.gov/programs-projects/external-rna-controls-consortium)."
      ],
      "metadata": {
        "id": "e5mNCok5HfhI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# define method for running PCA\n",
        "def run_pca(_expr, _meta):\n",
        "  # first only keep \"Ground Control\" and \"Space Flight\" samples\n",
        "  gc_samples = list(_meta[_meta['Factor Value[Spaceflight]'] == 'ground control']['Sample Name'])\n",
        "  sf_samples = list(_meta[_meta['Factor Value[Spaceflight]'] == 'space flight']['Sample Name'])\n",
        "  samples = gc_samples + sf_samples\n",
        "  expr = _expr[_expr['sample'].isin(samples)]\n",
        "  X = np.array(expr.drop(columns=['sample']))\n",
        "  print('X shape: ', X.shape)\n",
        "  factor_list = list()\n",
        "  for sample in list(expr['sample']):\n",
        "      factor_value = _meta[_meta['Sample Name'] == sample]['Factor Value[Spaceflight]'].values[0]\n",
        "      factor_list.append(factor_value)\n",
        "  y = np.array(list(factor_list))\n",
        "  print('y shape: ', y.shape)\n",
        "  target_names=np.array(['ground control', 'space flight'])\n",
        "\n",
        "  # run PCA to reduce dimensions from 23,419 to 2!\n",
        "  pca = PCA(n_components=2)\n",
        "  X_r = pca.fit_transform(X)\n",
        "\n",
        "  # Percentage of variance explained for each components\n",
        "  print(\n",
        "      \"explained variance ratio (first two components): %s\"\n",
        "      % str(pca.explained_variance_ratio_)\n",
        "  )\n",
        "\n",
        "  # plot the pca plot\n",
        "  plt.figure()\n",
        "  colors = [\"navy\", \"turquoise\"]\n",
        "  lw = 2\n",
        "  for color, i, target_name in zip(colors, ['ground control', 'space flight'], target_names):\n",
        "      plt.scatter(X_r[y == i, 0], X_r[y == i, 1], color=color, alpha=0.8, lw=lw, label=target_name)\n",
        "  plt.legend(loc=\"best\", shadow=False, scatterpoints=1)\n",
        "  plt.title(\"PCA of OSD dataset\")\n",
        "\n",
        "  # show the last digits of the sample name if the number of samples is less than 40\n",
        "  if len(expr) < 40:\n",
        "    samples = list(expr['sample'])\n",
        "    for i, txt in enumerate(samples):\n",
        "        plt.annotate(txt[-16:], (X_r[i][0], X_r[i][1]))\n",
        "  plt.show()\n"
      ],
      "metadata": {
        "id": "spcbjro1Vt9c"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# define method to convert ensembl gene id's to gene symbols\n",
        "def get_symbol_from_id(gene_id_list):\n",
        "  # Create an instance of EnsemblConverter\n",
        "  converter = EnsemblConverter()\n",
        "\n",
        "  # Convert Ensembl IDs to gene symbols\n",
        "  result = converter.convert_ids(gene_id_list)\n",
        "\n",
        "  # Print the resulting DataFrame\n",
        "  gene_symbol_list = list()\n",
        "  for i in range(len(result)):\n",
        "    gene_symbol_list.append(result.iloc[i]['Symbol'])\n",
        "\n",
        "  return gene_symbol_list"
      ],
      "metadata": {
        "id": "B_-3mrOpBvYo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Download data from OSDR"
      ],
      "metadata": {
        "id": "xybG83XEBR_w"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# open the URL printed below in another browser window\n",
        "url = !echo https://visualization.osdr.nasa.gov/biodata/api/v2/dataset/OSD-{OSD_DATASET}/files/\\?format=browser\n",
        "print(url[0])"
      ],
      "metadata": {
        "id": "XjmvscbU6emE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# download OSD unnormalized count data\n",
        "import pandas as pd\n",
        "url = 'https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + OSD_DATASET + '/download?source=datamanager&file=GLDS-' + GLDS_DATASET + '_rna_seq_Unnormalized_Counts.csv'\n",
        "# url = 'https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + OSD_DATASET + '/download?source=datamanager&file=GLDS-' + GLDS_DATASET + '_rna_seq_STAR_Unnormalized_Counts.csv'\n",
        "expr_df = pd.read_csv(url)\n",
        "expr_df.head()"
      ],
      "metadata": {
        "id": "1zKdaHOR6ssc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# identify the shape of the expr_df dataframe\n",
        "expr_df.shape"
      ],
      "metadata": {
        "id": "pAHP7a5j2oNJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# call drop_ercc\n",
        "# there may or may not be ERCC records to drop\n",
        "expr_df = drop_ercc(expr_df)\n",
        "expr_df.shape"
      ],
      "metadata": {
        "id": "KdvTv2CR3Fj3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# call drop_nans\n",
        "# there may or may not be NaNs to drop\n",
        "expr_df = drop_nans(expr_df)\n",
        "expr_df.shape"
      ],
      "metadata": {
        "id": "m9R771vtJlfY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# call drop_lowcount with threshold=50\n",
        "expr_df = drop_lowcount(expr_df, threshold=50)\n",
        "expr_df.shape"
      ],
      "metadata": {
        "id": "JbPKe_Y1IgTq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# call filter_cvs with thresh=CV_THRESH\n",
        "expr_df = filter_cvs(expr_df, thresh=CV_THRESH)\n",
        "expr_df.shape"
      ],
      "metadata": {
        "id": "fN5rK3g2IK0A"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# first save expr_df to file in gdrive for use in next notebook\n",
        "expr_df.to_csv(DGEA_DIR + '/expr_df.csv', sep=',', index=None)"
      ],
      "metadata": {
        "id": "CJuiDrboQyY_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# then transpose expr_df to be samples x genes\n",
        "expr_df = transpose_df(expr_df, 'Unnamed: 0', 'sample')\n",
        "expr_df.head()"
      ],
      "metadata": {
        "id": "FvCzUupYNX30"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# determine the shape of the expr_df_T dataframe\n",
        "expr_df.shape"
      ],
      "metadata": {
        "id": "8wcQSEJ_2w6a"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# read in  metadata\n",
        "url = 'https://osdr.nasa.gov/geode-py/ws/studies/OSD-' + OSD_DATASET + \\\n",
        "  '/download\\?source=datamanager\\&file=OSD-' + OSD_DATASET + '_metadata_OSD-' + GLDS_DATASET + '-ISA.zip'\n",
        "  # '/download\\?source=datamanager\\&file=OSD-' + OSD_DATASET + '_metadata_GLDS-' + GLDS_DATASET + '-ISA.zip'\n",
        "\n",
        "!curl -L {url} -o {META_DIR}/metadata-{OSD_DATASET}.zip\n",
        "!unzip -o {META_DIR}/metadata-{OSD_DATASET}.zip -d {META_DIR}"
      ],
      "metadata": {
        "id": "GS_GaujR7OF1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check metadata\n",
        "meta_df = pd.read_csv(f\"{META_DIR}/s_OSD-\" + OSD_DATASET + \".txt\", sep='\\t')\n",
        "meta_df.head()"
      ],
      "metadata": {
        "id": "1FQ_iTHrOQjw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# change Factor Value[Spaceflight] to lower case\n",
        "for i in range(len(meta_df)):\n",
        "  meta_df.loc[i, 'Factor Value[Spaceflight]'] = meta_df.iloc[i]['Factor Value[Spaceflight]'].lower().strip()"
      ],
      "metadata": {
        "id": "dLRSnssobX6v"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# see how many samples are flight and how many are ground\n",
        "meta_df['Factor Value[Spaceflight]'].value_counts()"
      ],
      "metadata": {
        "id": "Gc13wn26P3WC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# drop all values from expr_df and meta_df that are not either \"Ground Control\" or \"Space Flight\"\n",
        "gc_samples = list(meta_df[meta_df['Factor Value[Spaceflight]'] == 'ground control']['Sample Name'])\n",
        "sf_samples = list(meta_df[meta_df['Factor Value[Spaceflight]'] == 'space flight']['Sample Name'])\n",
        "samples = gc_samples + sf_samples\n",
        "expr_df = expr_df[expr_df['sample'].isin(samples)]\n",
        "meta_df = meta_df[meta_df['Sample Name'].isin(samples)]"
      ],
      "metadata": {
        "id": "9AP3RevTmzzB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check shape of final dataframes prior to analysis\n",
        "print('expr shape: ', expr_df.shape)\n",
        "print('meta shape: ', meta_df.shape)"
      ],
      "metadata": {
        "id": "14_wqaAV8N0j"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Run PCA\n",
        "\n",
        "Here we run principal components analysis (PCA) to determine if the samples separate in the plane."
      ],
      "metadata": {
        "id": "ojIgh_NfTIgu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# call run_pca on expr_df_final\n",
        "run_pca(expr_df, meta_df)"
      ],
      "metadata": {
        "id": "WE3Qbe4lWNuk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "As you can tell, the sample gene expression profiles separate very cleanly in the plane. All the flight samples are on the right and all the ground samples are on the left. The first principal component (horizontal axis) accounts for 89% of the variation in the data. The second principal component (vertical axis) accounts for 6% of the variation in the data. If we were to plot this in 3 dimensions, the third principal component would account for some percentage of the variation less than 6%."
      ],
      "metadata": {
        "id": "YMKy4XU-iGQz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Perform differential gene expression analysis"
      ],
      "metadata": {
        "id": "QlJK1EflNZEd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# First let's create a conditions table for 'Factor Value: Spaceflight'.\n",
        "# We'll assign 0 for ground control and 1 for spaceflight\n",
        "condition_dict=dict()\n",
        "for sample in list(expr_df['sample']):\n",
        "    val = meta_df[meta_df['Sample Name']==sample]['Factor Value[Spaceflight]'].values[0]\n",
        "    if val.lower().strip() == 'ground control':\n",
        "        condition_dict[sample] = 0\n",
        "    elif val.lower().strip() == 'space flight':\n",
        "        condition_dict[sample] = 1\n",
        "    else:\n",
        "      print('val ' + val + ' neither Ground Control nor Space Flight')\n",
        "conditions =  pd.DataFrame.from_dict(condition_dict, orient='index')\n",
        "conditions.columns = ['condition']\n"
      ],
      "metadata": {
        "id": "0TYCw7RqP1Q0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# examine conditions dataframe\n",
        "for i in range(len(meta_df)):\n",
        "  print(meta_df.iloc[i]['Sample Name'], meta_df.iloc[i]['Factor Value[Spaceflight]'], conditions.iloc[i]['condition'])"
      ],
      "metadata": {
        "id": "Dr6zQFfX5oNC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# make sure there are exactly 2 conditions for DGEA\n",
        "set(meta_df['Factor Value[Spaceflight]'])\n"
      ],
      "metadata": {
        "id": "byBB5Gn6YjQS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# create dataframe with 'sample' column removed for DESeq2\n",
        "import numpy as np\n",
        "counts = expr_df.drop(columns=['sample']).reset_index(drop=True)\n",
        "# check for non-numeric\n",
        "counts.applymap(np.isreal)\n",
        "# convert all cells to integers\n",
        "counts = counts.astype(int)\n",
        "counts.shape"
      ],
      "metadata": {
        "id": "_hTAn-Re2Ynr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can read more about the pyDeseq2 package in [this document](https://pydeseq2.readthedocs.io/en/stable/api/index.html#)."
      ],
      "metadata": {
        "id": "Y-9Bua7QSTfB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# run the DESEq2 experiment contrasting condition \"0\" (ground control) to condition \"1\" (space flight)\n",
        "# store the results in a dataframe called \"res\"\n",
        "dds = DeseqDataSet(counts=counts, metadata=conditions, design_factors=\"condition\")\n",
        "dds.deseq2()\n",
        "stats_results = DeseqStats(dds, contrast = ('condition', 0, 1))\n",
        "stats_results.summary()\n",
        "res = stats_results.results_df"
      ],
      "metadata": {
        "id": "Df_4q7H041KH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# extract results of the DESeq2 experiment where  padj < 0.05, and log2fc > LOG2FC_THRESH\n",
        "sigs = res[(res.padj < 0.05) & (abs(res.log2FoldChange) > LOG2FC_THRESH)]\n",
        "print(sigs.shape)\n",
        "sigs.head()\n",
        "sigs.to_csv(DGEA_DIR + '/sigs.csv', sep=',')"
      ],
      "metadata": {
        "id": "Uu5KVDFx8_8e"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# write differentially expressed gene list to file\n",
        "deg_genes = list(sigs.index)\n",
        "with open(f\"{DGEA_DIR}/deg_genes-\" + OSD_DATASET + \".txt\", \"w\") as f:\n",
        "  for gene in deg_genes:\n",
        "    f.write(gene + \"\\n\")\n",
        "f.close()\n",
        "\n",
        "# check gene file\n",
        "!head {DGEA_DIR}/deg_genes-{OSD_DATASET}.txt"
      ],
      "metadata": {
        "id": "iMLvVI83Dz-y"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# write list of background genes to file\n",
        "bg_genes = list(expr_df.drop(columns=['sample']))\n",
        "with open(f\"{DGEA_DIR}/background_genes-\" + OSD_DATASET + \".txt\", \"w\") as f:\n",
        "  for gene in bg_genes:\n",
        "    f.write(gene + \"\\n\")\n",
        "f.close()\n",
        "\n",
        "# check gene file\n",
        "!head {DGEA_DIR}/background_genes-{OSD_DATASET}.txt"
      ],
      "metadata": {
        "id": "bGbVDGmDlYgO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Create volcano plot\n",
        "\n",
        "You can learn more about how this volcano plot was generated reading [this article](https://bioinformatics.ccr.cancer.gov/docs/pies-2023/pies_lesson4/)."
      ],
      "metadata": {
        "id": "PUHaEOATD8xC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# import Python modules for plotting volcano\n",
        "import pandas\n",
        "import numpy\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn\n",
        "\n",
        "# create column for -log10PAdj from existing padj\n",
        "res['-log10PAdj'] = -numpy.log10(res['padj'])\n",
        "\n",
        "# create column for significance from existing log2FoldChange and -log10PAdj\n",
        "significance = []\n",
        "for i in range(len(res)):\n",
        "    if res.iloc[i]['log2FoldChange'] > LOG2FC_THRESH and res.iloc[i]['-log10PAdj'] > 5:\n",
        "        significance.append('significantly up-regulated')\n",
        "    elif res.iloc[i]['log2FoldChange'] < -LOG2FC_THRESH and res.iloc[i]['-log10PAdj'] > 5:\n",
        "        significance.append('significantly down-regulated')\n",
        "    elif res.iloc[i]['-log10PAdj'] > 5 and res.iloc[i]['padj'] < 0.05:\n",
        "        significance.append('significant')\n",
        "    else:\n",
        "        significance.append('not significant')\n",
        "\n",
        "res['significance'] = significance\n",
        "# plot the degs\n",
        "plot1 = seaborn.scatterplot(res, x = \"log2FoldChange\", y = \"-log10PAdj\", hue = \"significance\")\n",
        "\n",
        "# sort to get top 10 significant genes\n",
        "top10 = sigs[sigs['padj'] < 0.05].sort_values(by=['padj', 'log2FoldChange'], ascending=[True, False]).head(10)\n",
        "\n",
        "# add gene_symbol to top10\n",
        "gene_symbol_list = get_symbol_from_id(list(top10.index))\n",
        "top10['gene_symbol'] = gene_symbol_list\n",
        "\n",
        "# add -log10PAdj to top10\n",
        "top10['-log10PAdj'] = -numpy.log10(top10['padj'])\n",
        "\n",
        "# annotate top 10 significant genes\n",
        "for i in range(top10.shape[0]):\n",
        "    gene = top10.iloc[i]\n",
        "    plt.text(gene['log2FoldChange'], gene['-log10PAdj'], gene['gene_symbol'], fontsize=7, ha='right')\n",
        "\n",
        "# draw a line at y=5 for significance\n",
        "plt.axhline(y = 5, color='black', linestyle='--')\n",
        "\n",
        "# draw 2 lines at fold change threshold\n",
        "plt.axvline(x = LOG2FC_THRESH, color = 'black', linestyle='--')\n",
        "plt.axvline(x = -LOG2FC_THRESH, color='black', linestyle='--')\n",
        "\n",
        "# set font size for legend\n",
        "plt.legend(fontsize=8)\n"
      ],
      "metadata": {
        "id": "alTw2lzgEAWk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sigs"
      ],
      "metadata": {
        "id": "UKZmRkvsZfWV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The horizontal line (y=5) represents the negative log of the adjusted p-value (-log10PAdj). The value -log10(5), or 0.05, is the standard alpha level of significance for most scientific research. You can read more about the alpha level of significance [here](https://en.wikipedia.org/wiki/Statistical_significance).\n",
        "\n",
        "The two vertical lines (x=-LOG2FC_THRESH and x=+LOG2FC_THRESH) represent the number of times up- and down-regulated the gene expression value is between the 2 conditions of spaceflight and ground control, respectively. We defined the LOG2FC_THRESH value at the top of this notebook. You can read more about the log2 fold-change threshold for DGEA [here](https://www.biostars.org/p/9505003/).\n",
        "\n",
        "The list of differentially expressed genes include those which are significantly down-regulated (in green) and significantly up-regulated (in red)."
      ],
      "metadata": {
        "id": "-9qW6L0Ic97s"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Check your work before moving on"
      ],
      "metadata": {
        "id": "aIm5fVXzYRFk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# make sure your google drive disk space utilization is still under 15G\n",
        "# should be about 2.4GB\n",
        "!du -sh /content/mnt/MyDrive/NASA/GL4HS"
      ],
      "metadata": {
        "id": "2wVkwyoRYTec"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "end_time = datetime.datetime.now()\n",
        "print('notebook end time: ', end_time.strftime('%Y-%m-%d %H:%M:%S'))\n",
        "\n",
        "print('notebook runtime: ', end_time - start_time)\n",
        "#"
      ],
      "metadata": {
        "id": "Vq_JnOx_Mjkl"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}