{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/shreyansegnyte/NASA-GeneLab-Code/blob/main/preparingRawReads.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<div>\n",
        "<img src=\"https://www.nasa.gov/wp-content/uploads/2024/07/osdr-gl4hs-logo.png\" width=\"600\"/>\n",
        "</div>"
      ],
      "metadata": {
        "id": "w_4NkhWYaxyv"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **NOTEBOOK 2: Preparing RNA-seq reads**\n",
        "\n",
        "In this notebook, you will be preparing the RNA-seq reads you just downloaded in the previous notebook for subsequent processing. You will first calculate and compare the quality of each of the 2 FASTQ files. Next you will trim short, low-quality, and extraneous reads from each of the 2 FASTQ files. Last, you will calculate the quality again for one of the paired-end FASTQ files and compare it to what it was before trimming.\n",
        "\n",
        "## **Objectives of this notebook**\n",
        "The primary objective of this notebook is to ensure that the quality of the RNA-seq reads for our sample is sufficient.  To measure the quality of the FASTQ data, you will use the FastQC tool developed by Simon Andrews. You can learn more about FastQC in this [tutorial](https://rtsf.natsci.msu.edu/genomics/technical-documents/fastqc-tutorial-and-faq.aspx).\n",
        "\n",
        "You will install and use a tool called `trim_galore` to trim short reads, low-quality reads, and adapter reads from your FASTQ files. The `trim_galore` command is itself a wrapper of both the FastQC tool and the `cutadapt` tool. You can learn more about `trim_galore` in this [user guide](https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md).\n",
        "\n",
        "Note that there are a few ways you can display the images generated in this notebook. One way is to look at them inside the notebook. If they are too small to read, you can click on an image to enlarge it. Another way is to navigate inside your Google Drive folder, download the HTML/PDF/image files to your laptop, and look at them from your laptop.\n",
        "\n",
        "## **UNIX commands introduced in this notebook**\n",
        "\n",
        "[`apt-get`](https://manpages.ubuntu.com/manpages/lunar/man8/apt-get.8.html) command to install operating system packages from the Internet.\n",
        "\n",
        "[`pip`](https://pip.pypa.io/en/stable/cli/pip_install/) command to install Python packages from the Internet.\n",
        "\n",
        "[`tar`](https://man7.org/linux/man-pages/man1/tar.1.html) command to extract an archive of files.\n",
        "\n",
        "[`wget`](https://linux.die.net/man/1/wget) command to download files from the Internet.\n",
        "\n",
        "[`chmod`](https://man7.org/linux/man-pages/man1/chmod.1p.html) command to change permissions of a file.\n",
        "\n",
        "[`wkhtmltopdf`](https://wkhtmltopdf.org/usage/wkhtmltopdf.txt) to convert HTML to PDF.\n",
        "\n"
      ],
      "metadata": {
        "id": "OId01pUzOPdj"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prepare notebook environment"
      ],
      "metadata": {
        "id": "MbsdHhEvn0cc"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wuZaffD5o3mS"
      },
      "outputs": [],
      "source": [
        "# mount google drive\n",
        "from google.colab import drive\n",
        "drive.flush_and_unmount()\n",
        "drive.mount(\"mnt\")\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "start_time = datetime.datetime.now()\n",
        "print('notebook start time: ', start_time.strftime('%Y-%m-%d %H:%M:%S'))"
      ],
      "metadata": {
        "id": "9OwUBSa_zzKg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# make sure FASTQ_DIR directory exists on google drive\n",
        "import os\n",
        "FASTQ_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/FASTQ\"\n",
        "if not os.path.exists(FASTQ_DIR):\n",
        "  raise Exception('STOP! Make sure you run the previous notebook before running this one')\n"
      ],
      "metadata": {
        "id": "09QH8xP-yXcU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# make sure reduced_r1.fastq.gz and reduced_r2.fastq.gz files are in place\n",
        "if not os.path.exists(f\"{FASTQ_DIR}/reduced_r1.fastq.gz\") or not os.path.exists(f\"{FASTQ_DIR}/reduced_r2.fastq.gz\"):\n",
        "  raise Exception('STOP! Make sure you run the previous notebook before running this one')"
      ],
      "metadata": {
        "id": "Os6frKCRedon"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# create the TRIM directory for this notebook\n",
        "if os.path.exists(f\"{FASTQ_DIR}/TRIM\"):\n",
        "  !rm -rf {FASTQ_DIR}/TRIM\n",
        "!mkdir {FASTQ_DIR}/TRIM"
      ],
      "metadata": {
        "id": "-zLTvQgLPtf7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Run FastQC to check the quality of the FASTQ files before trimming"
      ],
      "metadata": {
        "id": "dUltvi3tvSJ6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# install FastQC\n",
        "if os.path.exists(f\"{FASTQ_DIR}/FastQC\"):\n",
        "  print('FastQC already installed - removing now to reinstall')\n",
        "  !rm -rf {FASTQ_DIR}/FastQC\n",
        "!mkdir {FASTQ_DIR}/FastQC\n",
        "!wget -O {FASTQ_DIR}/fastqc.zip https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.12.1.zip\n",
        "!unzip {FASTQ_DIR}/fastqc.zip -d {FASTQ_DIR} > /dev/null\n",
        "!chmod +x {FASTQ_DIR}/FastQC/fastqc"
      ],
      "metadata": {
        "id": "LrZnS5IyOA7y"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check the version of fastqc\n",
        "!{FASTQ_DIR}/FastQC/fastqc --version"
      ],
      "metadata": {
        "id": "SILqgtCpk8C5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# remove the fastqc.zip file as we don't need it anymore\n",
        "!rm -f {FASTQ_DIR}/fastqc.zip"
      ],
      "metadata": {
        "id": "0F8ARQWwif55"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# make directories for FASTQC output\n",
        "if os.path.exists(f\"{FASTQ_DIR}/FASTQC_R1_OUT\"):\n",
        "  !rm -rf {FASTQ_DIR}/FASTQC_R1_OUT\n",
        "!mkdir {FASTQ_DIR}/FASTQC_R1_OUT\n",
        "if os.path.exists(f\"{FASTQ_DIR}/FASTQC_R2_OUT\"):\n",
        "  !rm -rf {FASTQ_DIR}/FASTQC_R2_OUT\n",
        "!mkdir {FASTQ_DIR}/FASTQC_R2_OUT"
      ],
      "metadata": {
        "id": "TrPXJerlQqiE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# run fastqc on compressed reduced R1 fastq file\n",
        "!{FASTQ_DIR}/FastQC/fastqc  {FASTQ_DIR}/reduced_r1.fastq.gz -t 2 -o {FASTQ_DIR}/FASTQC_R1_OUT\n",
        "# check for error message"
      ],
      "metadata": {
        "id": "D87QKUIUAMgQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# list fastqc output directory\n",
        "# you should see the reduced R1 fastqc.html and fastqc.zip files that got generated in the previous step\n",
        "!ls -lh {FASTQ_DIR}/FASTQC_R1_OUT"
      ],
      "metadata": {
        "id": "SyzGtEEpA9Wp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# run fastqc on compressed reduced R2 fastq file\n",
        "!{FASTQ_DIR}/FastQC/fastqc  {FASTQ_DIR}/reduced_r2.fastq.gz -t 2 -o {FASTQ_DIR}/FASTQC_R2_OUT"
      ],
      "metadata": {
        "id": "z3x4IxXP4CoJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# list fastqc output directories\n",
        "# you should see the reduced R2 fastqc.html and fastqc.zip files that got generated in the previous step\n",
        "!ls -lh {FASTQ_DIR}/FASTQC_R2_OUT"
      ],
      "metadata": {
        "id": "jWhkCgy2CT5p"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# install packages to convert HTML to PDF for rendering in google colab\n",
        "!sudo DEBIAN_FRONTEND=noninteractive apt-get update -y > /dev/null\n",
        "!sudo DEBIAN_FRONTEND=noninteractive apt-get install -y wkhtmltopdf > /dev/null\n",
        "!sudo DEBIAN_FRONTEND=noninteractive apt-get install -y poppler-utils > /dev/null\n",
        "!pip install pdf2image > /dev/null\n"
      ],
      "metadata": {
        "id": "b_i3oCrhnI2P"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# run the wkhtmltopdf command to convert the FastQC HTML files into a PDF files\n",
        "!wkhtmltopdf {FASTQ_DIR}/FASTQC_R1_OUT/reduced_r1_fastqc.html {FASTQ_DIR}/FASTQC_R1_OUT/out.pdf\n",
        "!wkhtmltopdf {FASTQ_DIR}/FASTQC_R2_OUT/reduced_r2_fastqc.html {FASTQ_DIR}/FASTQC_R2_OUT/out.pdf"
      ],
      "metadata": {
        "id": "QdTKTjx4rhvn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# convert the PDF to images (there should be 6 images created from each PDF)\n",
        "from pdf2image import convert_from_path\n",
        "import os\n",
        "images_r1 = convert_from_path(f\"{FASTQ_DIR}/FASTQC_R1_OUT/out.pdf\")\n",
        "print('len r1 = ', str(len(images_r1)))\n",
        "images_r2 = convert_from_path(f\"{FASTQ_DIR}/FASTQC_R2_OUT/out.pdf\")\n",
        "print('len r2 = ', str(len(images_r2)))"
      ],
      "metadata": {
        "id": "kJZrV12PsyBx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# define a method to display a single page of the report\n",
        "\n",
        "def display_page(report_1, report_2, page_num):\n",
        "  from ctypes import resize\n",
        "  import matplotlib.pyplot as plt\n",
        "\n",
        "  plt.axis('off')\n",
        "  fig, axes = plt.subplots(1, 2, figsize=(30,30))\n",
        "  axes[0].imshow(report_1[page_num])\n",
        "  axes[0].axis('off')\n",
        "  axes[1].imshow(report_2[page_num])\n",
        "  axes[1].axis('off')"
      ],
      "metadata": {
        "id": "-rJLTWnC65g4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# display the first page of the report\n",
        "display_page(images_r1, images_r2, 0)\n"
      ],
      "metadata": {
        "id": "MRfJgR2Zb3Nd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Note the total number of sequences, total number of bases, number of sequences flagged as poor quality, sequence length, and percentage GC content for R1."
      ],
      "metadata": {
        "id": "eBk8Ovh_TTn6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# display the second page of the report\n",
        "display_page(images_r1, images_r2, 1)"
      ],
      "metadata": {
        "id": "XsLchSrcf5AY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# display the third page of the report\n",
        "display_page(images_r1, images_r2, 2)"
      ],
      "metadata": {
        "id": "lJSG2ydsZpPp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# display the fourth page of the report\n",
        "display_page(images_r1, images_r2, 3)"
      ],
      "metadata": {
        "id": "61qaS7jKiPMy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# display the fifth page of the report\n",
        "display_page(images_r1, images_r2, 4)"
      ],
      "metadata": {
        "id": "LQGdZjD3iUHi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "# Use `trim_galore` to trim the reads"
      ],
      "metadata": {
        "id": "hgR8448WusIq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install Trim Galore\n",
        "import os\n",
        "if os.path.exists(f\"{FASTQ_DIR}/TRIM/TrimGalore-0.6.10/trim_galore\"):\n",
        "  print('trim_galore already installed - removing now to reinstall')\n",
        "  !rm -rf {FASTQ_DIR}/TRIM\n",
        "!mkdir {FASTQ_DIR}/TRIM\n",
        "!wget -O {FASTQ_DIR}/TRIM/trim_galore.tar.gz https://github.com/FelixKrueger/TrimGalore/archive/0.6.10.tar.gz\n",
        "!curl -fsSL https://github.com/FelixKrueger/TrimGalore/archive/0.6.10.tar.gz -o {FASTQ_DIR}/TRIM/trim_galore.tar.gz\n",
        "!tar xzf {FASTQ_DIR}/TRIM/trim_galore.tar.gz -C {FASTQ_DIR}/TRIM\n",
        "# make the trim_galore command executable\n",
        "!chmod +x {FASTQ_DIR}/TRIM/TrimGalore-0.6.10/trim_galore\n",
        "# remove the compressed tar file\n",
        "!rm -f {FASTQ_DIR}/TRIM/trim_galore.tar.gz"
      ],
      "metadata": {
        "id": "DfLtsXfFDydr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check version of trim_galore\n",
        "!{FASTQ_DIR}/TRIM/TrimGalore-0.6.10/trim_galore -v"
      ],
      "metadata": {
        "id": "fOahtC6mkOI3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# install cutadapt\n",
        "!pip install cutadapt"
      ],
      "metadata": {
        "id": "jMx72vATcydp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# find path to cutadapt executable\n",
        "!which cutadapt"
      ],
      "metadata": {
        "id": "CdZotVwzc2Du"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# run trim_galore on R1 and R2\n",
        "# use the path_to_cutadapt path you found in the previous code cell\n",
        "if os.path.exists(f\"{FASTQ_DIR}/TRIM/PAIRED\"):\n",
        "  !rm -rf {FASTQ_DIR}/TRIM/PAIRED\n",
        "!mkdir -p {FASTQ_DIR}/TRIM/PAIRED\n",
        "\n",
        "!{FASTQ_DIR}/TRIM/TrimGalore-0.6.10/trim_galore \\\n",
        "  --path_to_cutadapt /usr/local/bin/cutadapt \\\n",
        "  --paired \\\n",
        "  -o {FASTQ_DIR}/TRIM/PAIRED \\\n",
        "  -q 20 \\\n",
        "  -j 2 \\\n",
        "  {FASTQ_DIR}/reduced_r1.fastq.gz \\\n",
        "  {FASTQ_DIR}/reduced_r2.fastq.gz\n"
      ],
      "metadata": {
        "id": "ieyMS7U1HFiC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Note the following:\n",
        "1. which adapter sequence was auto-detected and most prevalent?\n",
        "2. what is the minimum required sequence length for both reads before a sequence pair gets removed?\n",
        "3. approximately what percentage of sequence pairs were removed because at least one read was shorter than the length cutoff?\n",
        "4. which base was the most prevalent preceding removed adapters?"
      ],
      "metadata": {
        "id": "kYLU9vasOQa1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# validate the trimmed output files got created\n",
        "# should be about 2.1GB each\n",
        "!ls -lh {FASTQ_DIR}/TRIM/PAIRED\n"
      ],
      "metadata": {
        "id": "ItdTc4XQe5DP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# examine trimming report (or refer student cell with trim command that also has output)\n",
        "#!cat {FASTQ_DIR}/TRIM/R1/reduced_r1.fastq.gz_trimming_report.txt\n",
        "#!cat {FASTQ_DIR}/TRIM/R2/reduced_r2.fastq.gz_trimming_report.txt"
      ],
      "metadata": {
        "id": "0U9wDkbvdlcM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Run FastQC to check the quality of FASTQ file after trimming\n",
        "\n",
        "In this section, the images we display on the left will be the pre-trimming images for R1 only and the images on the right will be the corresponding post-trimming images for R1 only. We will not compare pre- and post-trimming reports for R2."
      ],
      "metadata": {
        "id": "pAQtTs6Xj4dJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# repeat FastQC run and examine quality scores across all bases + quality score distribution for R1\n",
        "# TODO do some sort of before-and-after side-by-side comparison of plots\n",
        "if os.path.exists(f\"{FASTQ_DIR}/TRIM/R1/FASTQC_OUT\"):\n",
        "  !rm -rf {FASTQ_DIR}/TRIM/R1/FASTQC_OUT\n",
        "!mkdir -p {FASTQ_DIR}/TRIM/R1/FASTQC_OUT\n",
        "!{FASTQ_DIR}/FastQC/fastqc  {FASTQ_DIR}/TRIM/PAIRED/reduced_r1_val_1.fq.gz -o {FASTQ_DIR}/TRIM/R1/FASTQC_OUT"
      ],
      "metadata": {
        "id": "MMtKLG7Bzvwj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# convert HTML to pdf\n",
        "!wkhtmltopdf {FASTQ_DIR}/TRIM/R1/FASTQC_OUT/reduced_r1_val_1_fastqc.html {FASTQ_DIR}/TRIM/R1/FASTQC_OUT/out.pdf"
      ],
      "metadata": {
        "id": "nqwrVCqUfdYv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# convert PDF to pillow image\n",
        "from pdf2image import convert_from_path\n",
        "images_r1_trim = convert_from_path(f\"{FASTQ_DIR}/TRIM/R1/FASTQC_OUT/out.pdf\")\n",
        "print('len = ', str(len(images_r1_trim)))"
      ],
      "metadata": {
        "id": "RykAJojugCz-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# compare first page of FastQC output for pre- and post-trim R1 data\n",
        "display_page(images_r1, images_r1_trim, 0)"
      ],
      "metadata": {
        "id": "-Smzaq0FgaOt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# compare second page of FastQC output for pre- and post-trim R1 data\n",
        "display_page(images_r1, images_r1_trim, 1)"
      ],
      "metadata": {
        "id": "tWpg1r3bgePD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# compare third page of FastQC output for pre- and post-trim R1 data\n",
        "display_page(images_r1, images_r1_trim, 2)"
      ],
      "metadata": {
        "id": "FYAZ7sP7gxeF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# compare fourth page of FastQC output for pre- and post-trim R1 data\n",
        "display_page(images_r1, images_r1_trim, 3)"
      ],
      "metadata": {
        "id": "b47mByIwg4Z2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# compare fifth page of FastQC output for pre- and post-trim R1 data\n",
        "display_page(images_r1, images_r1_trim, 4)"
      ],
      "metadata": {
        "id": "nvQEPdu1g8WX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# compare sixth page of FastQC output for pre- and post-trim R1 data\n",
        "display_page(images_r1, images_r1_trim, 5)"
      ],
      "metadata": {
        "id": "Gzp6hIDyhA_e"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Check your work before moving on"
      ],
      "metadata": {
        "id": "vsbDHoqIzceh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# check size of all GL4HS drive usage\n",
        "# should be about 260MB\n",
        "!du -sh /content/mnt/MyDrive/NASA/GL4HS"
      ],
      "metadata": {
        "id": "YpdUOpWczhSE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "end_time = datetime.datetime.now()\n",
        "print('notebook end time: ', end_time.strftime('%Y-%m-%d %H:%M:%S'))\n",
        "\n",
        "total_notebook_time = end_time - start_time\n",
        "print('total notebook time: ', total_notebook_time)"
      ],
      "metadata": {
        "id": "cYUs4pqOz1pP"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}