{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/shreyansegnyte/NASA-GeneLab-Code/blob/main/read%20alignment.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<div>\n",
        "<img src=\"https://www.nasa.gov/wp-content/uploads/2024/07/osdr-gl4hs-logo.png\" width=\"600\"/>\n",
        "</div>"
      ],
      "metadata": {
        "id": "5zwEdZQWa8dw"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **NOTEBOOK 4: Aligning sequencing reads to a reference genome**\n",
        "\n",
        "\n",
        "In this notebook you will map (align) the sequence reads from the FASTQ files to the reference chromosome index (chr17) you built in the previous notebook.\n",
        "\n",
        "## **Objectives of this notebook**\n",
        "The primary objective of this notebook is to align the 2 FASTQ files to a reference genome index for chromosome 17. Once you perform the alignment, you will check the alignment using a command called `samtools`. You can learn more about the `samtools` command in this [Wikipedia article](https://en.wikipedia.org/wiki/SAMtools).\n",
        "\n",
        "## **UNIX commands introduced in this notebook**\n"
      ],
      "metadata": {
        "id": "yvasx_p7Ykbu"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prepare your runtime environment for this lab"
      ],
      "metadata": {
        "id": "RMXtsrPS3pFs"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wuZaffD5o3mS"
      },
      "outputs": [],
      "source": [
        "# mount the google drive\n",
        "from google.colab import drive\n",
        "drive.flush_and_unmount()\n",
        "drive.mount(\"mnt\")\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "start_time = datetime.datetime.now()\n",
        "print('notebook start time: ', start_time.strftime('%Y-%m-%d %H:%M:%S'))"
      ],
      "metadata": {
        "id": "sZ1_qzKH8jCg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# import the os module which you'll use throughout the notebook\n",
        "import os"
      ],
      "metadata": {
        "id": "zTH9Flk46nLA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set FASTQ_DIR and verify it exists\n",
        "FASTQ_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/FASTQ\"\n",
        "if not os.path.exists(FASTQ_DIR):\n",
        "  raise Exception(\"STOP! You have not finished the previous notebooks!\")"
      ],
      "metadata": {
        "id": "hKTKRCHk6fM4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set REFERENCE_DIR and verify it exists\n",
        "REFERENCE_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/REFERENCE\"\n",
        "if not os.path.exists(REFERENCE_DIR):\n",
        "  raise Exception(\"STOP! You have not finished the previous notebooks!\")"
      ],
      "metadata": {
        "id": "EMywa4uq6PXY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# set STAR_DIR and verify it exists\n",
        "STAR_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/STAR\"\n",
        "if not os.path.exists(STAR_DIR):\n",
        "  raise Exception(\"STOP! You have not finished the previous notebooks!\")"
      ],
      "metadata": {
        "id": "oSJSINkR6XAP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check for the compressed trimmed fastq files\n",
        "if not os.path.exists(f\"{FASTQ_DIR}/TRIM/PAIRED/reduced_r1_val_1.fq.gz\") \\\n",
        "  or not os.path.exists(f\"{FASTQ_DIR}/TRIM/PAIRED/reduced_r2_val_2.fq.gz\"):\n",
        "  raise Exception(\"STOP: one or both of the reduced trimmed fq.gz files do not exist. You must first run the previous notebooks!\")"
      ],
      "metadata": {
        "id": "it99dsPh33AO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# install samtools\n",
        "!sudo apt-get install samtools > /dev/null 2>&1"
      ],
      "metadata": {
        "id": "zZyzMBn96xuF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check version of samtools\n",
        "!samtools --version"
      ],
      "metadata": {
        "id": "WWoKmdcKBzxn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Align the reads to the reference"
      ],
      "metadata": {
        "id": "DDG-EQW03upW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# unzip the reduce_trimmed fq.gz files as STAR only reads unzipped fastq\n",
        "#\n",
        "import os\n",
        "if os.path.exists(f\"{FASTQ_DIR}/TRIM/PAIRED/reduced_r1_val_1.fq.gz\") and not os.path.exists(f\"{FASTQ_DIR}/TRIM/PAIRED/reduced_r1_val_1.fq\"):\n",
        "  !gunzip -f {FASTQ_DIR}/TRIM/PAIRED/reduced_r1_val_1.fq.gz\n",
        "if os.path.exists(f\"{FASTQ_DIR}/TRIM/PAIRED/reduced_r2_val_2.fq.gz\") and not os.path.exists(f\"{FASTQ_DIR}/TRIM/PAIRED/reduced_r2_val_2.fq\"):\n",
        "  !gunzip -f {FASTQ_DIR}/TRIM/PAIRED/reduced_r2_val_2.fq.gz"
      ],
      "metadata": {
        "id": "p285Z0WQqIkg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Use the `STAR` command to perform the alignment"
      ],
      "metadata": {
        "id": "5jiAilffyWz5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read Sections 3.1 and 3.2 of the [STAR manual](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf) to learn about the STAR command options for running mapping (alignment) jobs.\n",
        "\n",
        "For more information, feel free to read [this tutorial](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/03_alignment.html) on the STAR aligner for more information about how it works."
      ],
      "metadata": {
        "id": "4pyvIsnfyh7b"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# run STAR to align the reads to the reference\n",
        "# this may take a LONG time (several hours)\n",
        "# runtime depends on the REDUCTION_FACTOR setting from the first notebook\n",
        "# runtime also depends on how much data you're aligning\n",
        "# runtime also depends on how much memory is on the computer where STAR is running\n",
        "import datetime\n",
        "start = datetime.datetime.now()\n",
        "print('starting time: ', start.strftime('%Y-%m-%d %H:%M:%S'))\n",
        "\n",
        "!chmod +x {STAR_DIR}/STAR-2.7.11b/bin/Linux_x86_64_static/STAR\n",
        "\n",
        "if os.path.exists(f\"{STAR_DIR}/ALIGNMENT\"):\n",
        "  !rm -rf {STAR_DIR}/ALIGNMENT\n",
        "!mkdir {STAR_DIR}/ALIGNMENT\n",
        "!{STAR_DIR}/STAR-2.7.11b/bin/Linux_x86_64_static/STAR \\\n",
        "  --outFileNamePrefix {STAR_DIR}/ALIGNMENT/chr17 \\\n",
        "  --readFilesIn {FASTQ_DIR}/TRIM/PAIRED/reduced_r1_val_1.fq {FASTQ_DIR}/TRIM/PAIRED/reduced_r2_val_2.fq \\\n",
        "  --genomeDir {REFERENCE_DIR}/MM39_CHR17 \\\n",
        "  --runThreadN 2\n",
        "\n",
        "end = datetime.datetime.now()\n",
        "print('ending time: ', end.strftime('%Y-%m-%d %H:%M:%S'))\n"
      ],
      "metadata": {
        "id": "09QH8xP-yXcU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check alignment output files\n",
        "# there should be 5 files in all\n",
        "# the alignment file itself is called chr17Aligned.out.sam\n",
        "!ls -lh {STAR_DIR}/ALIGNMENT/chr17*"
      ],
      "metadata": {
        "id": "1y-YluDyC2bn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Use `samtools` view to examine the alignment"
      ],
      "metadata": {
        "id": "y8krb_iYxoOa"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read the [samtools view manpage](https://www.htslib.org/doc/samtools-view.html) to learn more about this command."
      ],
      "metadata": {
        "id": "psU83eqIxToj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# look at the first 10 lines of the SAM file\n",
        "# Question: Which position is the first read in this (unsorted) SAM file?\n",
        "!samtools view -h {STAR_DIR}/ALIGNMENT/chr17Aligned.out.sam | head -10"
      ],
      "metadata": {
        "id": "Zy4_OkM-Ni3j"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# sort alignment and save in BAM file\n",
        "!samtools sort {STAR_DIR}/ALIGNMENT/chr17Aligned.out.sam -o {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam"
      ],
      "metadata": {
        "id": "nYW7mmXqHfzd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# look at the first 10 lines of the sorted BAM file\n",
        "# Question: Which position is the first read in this (sorted) BAM file?\n",
        "!samtools view -h {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam | head -10"
      ],
      "metadata": {
        "id": "gZLmW3iv9cN-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# count all aligned reads in BAM file\n",
        "!samtools view -c {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam"
      ],
      "metadata": {
        "id": "2I4Igi2zCCYA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# count number of reads with MAPQ quality score 20 or higher in BAM file\n",
        "!samtools view -q 20 -c {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam"
      ],
      "metadata": {
        "id": "Q46pOUDqCmrX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# calculate percentage of reads with MAPQ > 20 in BAM file\n",
        "q20 = !samtools view -q 20 -c {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam\n",
        "total = !samtools view -c {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam\n",
        "int(q20[0]) / int(total[0])"
      ],
      "metadata": {
        "id": "9AGM9HBcDEwi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Use `samtools` flagstat to get statistics on the alignment"
      ],
      "metadata": {
        "id": "RSerIDe0x3Gh"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read the [samtools flagstat manpage](https://www.htslib.org/doc/samtools-flagstat.html) for more information."
      ],
      "metadata": {
        "id": "77hMDQH4x-Q5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# get general stats on alignment from BAM file\n",
        "!samtools flagstat {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam"
      ],
      "metadata": {
        "id": "L_R-qxCHGGYT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Singletons are reads that are missing a mate (R1 or R2) in paired-end sequencing. You can read more [here](https://www.seqanswers.com/forum/bioinformatics/bioinformatics-aa/41311-what-is-singletons)."
      ],
      "metadata": {
        "id": "5z-1uxACPK_r"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Use the `samtools mpileup` command to examine individual base mappings"
      ],
      "metadata": {
        "id": "wVwqbJtNyJ_b"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can read more about samtools mpileup in the DESCRIPTION Pileup Format section of the [samtools manpage](https://www.htslib.org/doc/samtools-mpileup.html) and in this [Wikipedia article](https://en.wikipedia.org/wiki/Pileup_format)."
      ],
      "metadata": {
        "id": "LeCFjl1uQOgT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# run mpileup in a region of the alignment in BAM file\n",
        "!samtools mpileup -f {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa {STAR_DIR}/ALIGNMENT/chr17Aligned.out.bam | sed -n '1,30p'"
      ],
      "metadata": {
        "id": "2tRv4-6xG1q3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Check your work before moving on"
      ],
      "metadata": {
        "id": "Ym1NH3ZK9CvH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# check disk space utilization (should be about 1.2GB)\n",
        "!du -sh /content/mnt/MyDrive/NASA/GL4HS"
      ],
      "metadata": {
        "id": "KbF6jOeh9Fpe"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "end_time = datetime.datetime.now()\n",
        "print('notebook end time: ', end_time.strftime('%Y-%m-%d %H:%M:%S'))\n",
        "\n",
        "total_time = end_time - start_time\n",
        "print('notebook total runtime: ', total_time)"
      ],
      "metadata": {
        "id": "GkvTo2nh8rvB"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}