{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/shreyansegnyte/NASA-GeneLab-Code/blob/main/refgenome.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<div>\n",
        "<img src=\"https://www.nasa.gov/wp-content/uploads/2024/07/osdr-gl4hs-logo.png\" width=\"600\"/>\n",
        "</div>"
      ],
      "metadata": {
        "id": "1fH4Ip6za3W6"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **NOTEBOOK 3: Building a reference genome index**\n",
        "\n",
        "\n",
        "In this notebook you will build a reference chromosome index (chr17) for hands-on experience and demonstration purposes.\n",
        "\n",
        "## **Objectives of this notebook**\n",
        "The primary objective of this notebook is to build a reference genome index for chromosome 17. A reference genome index (like any index) makes searching faster, so the overall performance of the mapping/alignment step will also be faster. In subsequent notebooks, you will use this index to align the sequence records from the FASTQ files to this reference genome. You can learn more about reference genomes in this [Wikipedia article](https://en.wikipedia.org/wiki/Reference_genome).\n",
        "\n",
        "## **UNIX commands introduced in this notebook**\n",
        "\n",
        "[`tail`](https://man7.org/linux/man-pages/man1/tail.1.html) command to see the last n lines of a file.\n",
        "\n",
        "[`mkdir`](https://man7.org/linux/man-pages/man1/mkdir.1.html) command to make a directory."
      ],
      "metadata": {
        "id": "rp15xWumWZlI"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prepare runtime environment"
      ],
      "metadata": {
        "id": "GtT-4qvPuPfV"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wuZaffD5o3mS"
      },
      "outputs": [],
      "source": [
        "# mount google drive for notebook\n",
        "from google.colab import drive\n",
        "drive.flush_and_unmount()\n",
        "drive.mount(\"mnt\")\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "start_time = datetime.datetime.now()\n",
        "print('notebook start time: ', start_time.strftime('%Y-%m-%d %H:%M:%S'))"
      ],
      "metadata": {
        "id": "S4HtN0KX1o1p"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# define FASTQ_DIR\n",
        "import os\n",
        "FASTQ_DIR=\"/content/mnt/MyDrive/NASA/GL4HS/FASTQ\"\n",
        "if not os.path.exists(FASTQ_DIR):\n",
        "  raise Exception(\"STOP! You haven't completed the previous notebooks yet\")"
      ],
      "metadata": {
        "id": "09QH8xP-yXcU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# create directory structure for this lab\n",
        "import os\n",
        "REFERENCE_DIR='/content/mnt/MyDrive/NASA/GL4HS/REFERENCE'\n",
        "if os.path.exists(REFERENCE_DIR):\n",
        "  !rm -rf {REFERENCE_DIR}\n",
        "!mkdir {REFERENCE_DIR}\n"
      ],
      "metadata": {
        "id": "BxvgWvXzj-21"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# create directory structure for this lab\n",
        "import os\n",
        "STAR_DIR='/content/mnt/MyDrive/NASA/GL4HS/STAR'\n",
        "if os.path.exists(STAR_DIR):\n",
        "  !rm -rf {STAR_DIR}\n",
        "!mkdir {STAR_DIR}"
      ],
      "metadata": {
        "id": "ctXlum7gwM8D"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# download and install STAR\n",
        "if not os.path.exists('/content/mnt/MyDrive/NASA/GL4HS/STAR/bin/Linux_x86_64_static/STAR'):\n",
        "  !wget -O {STAR_DIR}/STAR.tar.gz https://github.com/alexdobin/STAR/archive/2.7.11b.tar.gz\n",
        "  !tar -xzf {STAR_DIR}/STAR.tar.gz -C {STAR_DIR}\n",
        "\n",
        "!chmod +x {STAR_DIR}/STAR-2.7.11b/bin/Linux_x86_64_static/STAR\n",
        "\n",
        "# remove the compressed tar file\n",
        "!rm {STAR_DIR}/STAR.tar.gz"
      ],
      "metadata": {
        "id": "mnCJAdWcjyMW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check version of STAR\n",
        "!{STAR_DIR}/STAR-2.7.11b/bin/Linux_x86_64_static/STAR --version"
      ],
      "metadata": {
        "id": "XFDqrvAWwuwt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# download GRCm39 reference for chromosome 17\n",
        "import os\n",
        "if not os.path.exists(f\"{REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa.gz\"):\n",
        "  !wget -O {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa.gz https://ftp.ensembl.org/pub/release-113/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.chromosome.17.fa.gz\n",
        "  !gunzip -c {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa.gz > {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa\n"
      ],
      "metadata": {
        "id": "AG7HwbZhj05d"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# look at the first 10 lines of the reference fasta file\n",
        "!head {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa"
      ],
      "metadata": {
        "id": "JxTZBVBEMCEb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# look at 10 lines in the middle of the reference fasta file\n",
        "!sed -n '100000,100010 p' {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa"
      ],
      "metadata": {
        "id": "5UlcDnm5vOtF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# look at the last 10 lines of the reference fasta file\n",
        "!tail {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa"
      ],
      "metadata": {
        "id": "6LJM-zERx3MM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read [this discussion thread](https://www.reddit.com/r/genetics/comments/rz47pq/why_there_is_a_lot_of_ns_at_the_begining_of_the/) to see why there may be lots of 'N' at beginning and end of a reference genome fasta file."
      ],
      "metadata": {
        "id": "zFibWMBJzH1M"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Run STAR to build reference chromosome index"
      ],
      "metadata": {
        "id": "Ksqq7egwuezN"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read [this discussion thread](https://www.biostars.org/p/251736/) and search the [documentation](https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf) to learn more about `genomeChrBinNbits` option of the STAR command."
      ],
      "metadata": {
        "id": "cw4T2r1GIY7o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# run STAR to create index of GRCm39 chr 17 reference\n",
        "if os.path.exists(REFERENCE_DIR + '/MM39_CHR17'):\n",
        "  !rm -rf {REFERENCE_DIR}/MM39_CHR17\n",
        "!mkdir -p {REFERENCE_DIR}/MM39_CHR17\n",
        "!{STAR_DIR}/STAR-2.7.11b/bin/Linux_x86_64_static/STAR \\\n",
        "        --runThreadN 2 \\\n",
        "        --runMode genomeGenerate \\\n",
        "        --genomeDir {REFERENCE_DIR}/MM39_CHR17 \\\n",
        "        --genomeFastaFiles {REFERENCE_DIR}/Mus_musculus.GRCm39.dna.chromosome.17.fa  \\\n",
        "        --genomeSAindexNbases 12 \\\n",
        "        --genomeChrBinNbits 5"
      ],
      "metadata": {
        "id": "7EmmJhFWr_sk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check index file\n",
        "!ls -lh {REFERENCE_DIR}/MM39_CHR17"
      ],
      "metadata": {
        "id": "breihqoWuyoA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# check size of google drive usage for the reference (should be about 1.1G)\n",
        "!du -sh {REFERENCE_DIR}/MM39_CHR17"
      ],
      "metadata": {
        "id": "h3hcejRnzDZ6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Check your work before moving on"
      ],
      "metadata": {
        "id": "Bi38ArXrzs-A"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# check size of all GL4HS drive usage (should be about 1.4G)\n",
        "!du -sh /content/mnt/MyDrive/NASA/GL4HS"
      ],
      "metadata": {
        "id": "hOl7Z-owzMmR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# time the notebook\n",
        "import datetime\n",
        "end_time = datetime.datetime.now()\n",
        "print('notebook end time: ', end_time.strftime('%Y-%m-%d %H:%M:%S'))\n",
        "\n",
        "total_time = end_time - start_time\n",
        "print('notebook total runtime: ', total_time)"
      ],
      "metadata": {
        "id": "XXlbSWuL1yX4"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}